samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -462,6 +436,39 @@ object SamzaContainer extends Logging {
     else
       null
 
+    def maybeAsyncStreamTaskFactory : Object = {
+
+      val taskClassFactory = taskFactory match {
+        case null => TaskFactories.fromTaskClassConfig(config)
+        case _ => taskFactory
+      }
+
+      if (taskClassFactory == null) {
+        throw new SamzaException("Either the task class name or the task factory instance is required.")
+      }
+
+      val isAsyncTaskClass: Boolean = taskClassFactory.isInstanceOf[AsyncStreamTaskFactory]
+
+      if (isAsyncTaskClass) {
+        info("Got an AsyncStreamTask implementation.")
+      }
+
+      if (singleThreadMode && isAsyncTaskClass) {
+        throw new SamzaException("AsyncStreamTask cannot run on single thread mode.")
+      }
+
+      if (!singleThreadMode && !isAsyncTaskClass) {
+        info("Converting StreamTask to AsyncStreamTaskAdapter when running StreamTask w/ multiple threads")
+        new AsyncStreamTaskFactory {
```

Comment:
I tried. The difficulty is mainly in the taskThreadPool: it is instantiated within SamzaContainer context and optionally used in this case. The same taskThreadPool is needed outside the TaskFactories as well, which makes it really messy. There is an option to passing all these parameters to TaskFactories.fromConfig(). Let me make one more try to see whether the code looks better.